### PR TITLE
 Fix bug in result of SolverCG when run in interleaved mode

### PR DESCRIPTION
**Pull request by kronbichler**
This fixes a bug originally introduced by #13698, but then (erroneously) addressed by #13723. The problem was due to the eager update formula. To reduce the data access, we update the result vector only every other iteration. In case `SolverCG` terminates in an even number of iterations, we then needed to do the skipped update step. What happened was that we simply copied the update formula from https://github.com/dealii/dealii/blob/92288b9f2268036e468786f4bcb3f96d820813b1/include/deal.II/lac/solver_cg.h#L887-L889 with constants
https://github.com/dealii/dealii/blob/92288b9f2268036e468786f4bcb3f96d820813b1/include/deal.II/lac/solver_cg.h#L857-L860 to the place https://github.com/dealii/dealii/blob/92288b9f2268036e468786f4bcb3f96d820813b1/include/deal.II/lac/solver_cg.h#L970-L980
forgetting about the fact that we did not perform the switch `beta` -> `previous_beta` yet at the stage we terminate the code. To ensure that things are right, I added a test that terminates at all possible numbers of iterations between 1 and 11 and printing the result vector.

### CODE REVIEW

**Comment by kronbichler** [comment]
*ping*

File: include/deal.II/lac/solver_cg.h [hunk 1]
```diff
@@ -852,6 +852,38 @@ namespace internal
...
+                pj = beta * pj + prec_rj;
+                pj.store(p + j);
+                rj = VectorizedArray<Number>();
+                rj.store(v + j);
+              }
+            for (unsigned int j = end_regular; j < end_range; ++j)
+              {
+                r[j] -= alpha * v[j];
+                x[j] += alpha * p[j];
+                p[j] = beta * p[j] + this->preconditioner.apply(j, r[j]);
```
Comment (masterleinad) [inline include/deal.II/lac/solver_cg.h:883]:
```suggestion
                p[j] = this->preconditioner.apply(j, r[j]);
```

File: include/deal.II/lac/solver_cg.h [hunk 2]
```diff
@@ -1042,6 +1074,32 @@ namespace internal
...
+                for (unsigned int i = 0; i < length; ++i)
+                  {
+                    p[i] = this->beta * p[i] + prec_r[i];
+                    v[i] = Number();
+                  }
+                p += length;
+                r += length;
+                v += length;
+              }
+          }
```
Comment (masterleinad) [inline include/deal.II/lac/solver_cg.h:1102]: Isn't this the exact same code as for `iteration_index % 2 == 0 && beta != Number()` or did you mean to exploit `beta==0` in line 1095?

File: include/deal.II/lac/solver_cg.h [hunk 3]
```diff
@@ -1042,6 +1074,32 @@ namespace internal
...
+                for (unsigned int i = 0; i < length; ++i)
+                  x[i] += this->alpha * p[i];
+                this->preconditioner.apply_to_subrange(j,
+                                                       j + length,
+                                                       r,
+                                                       prec_r.data());
+                DEAL_II_OPENMP_SIMD_PRAGMA
+                for (unsigned int i = 0; i < length; ++i)
+                  {
+                    p[i] = this->beta * p[i] + prec_r[i];
```
Comment (masterleinad) [inline include/deal.II/lac/solver_cg.h:1095]:
```suggestion
                    p[i] = prec_r[i];
```

File: include/deal.II/lac/solver_cg.h (hunk 2 above)
Comment (kronbichler) [reply include/deal.II/lac/solver_cg.h:1102]: It is almost the same, we add the line to compute `x` (line 1087). I add the comment from above and restructure the code slightly.

**Review by masterleinad — APPROVED** [review]

**Pull request merged**